pkg/commands/operation/operation.go

```
@@ -78,41 +79,36 @@ func DownloadVEXRepositories(ctx context.Context, opts flag.Options) error {
 }
 
 // InitBuiltinChecks downloads the built-in policies and loads them
-func InitBuiltinChecks(ctx context.Context, cacheDir string, quiet, skipUpdate bool, checkBundleRepository string, registryOpts ftypes.RegistryOptions) ([]string, error) {
+func InitBuiltinChecks(ctx context.Context, client *policy.Client, skipUpdate bool, registryOpts ftypes.RegistryOptions) (string, error) {
 	mu.Lock()
 	defer mu.Unlock()
+	var err error
 
-	client, err := policy.NewClient(cacheDir, quiet, checkBundleRepository)
-	if err != nil {
-		return nil, xerrors.Errorf("check client error: %w", err)
-	}
+	if skipUpdate {
+		log.Info("No downloadable checks were loaded as --skip-check-update is enabled, loading from existing cache...")
 
-	needsUpdate := false
-	if !skipUpdate {
-		needsUpdate, err = client.NeedsUpdate(ctx, registryOpts)
+		path := client.LoadBuiltinChecks()
+		_, _, err := misconf.CheckPathExists(path)
 		if err != nil {
-			return nil, xerrors.Errorf("unable to check if built-in policies need to be updated: %w", err)
+			return "", xerrors.Errorf("Failed to load existing cache, err: %s falling back to embedded checks...", err.Error())
 		}
+		return path, nil
+	}
+
+	needsUpdate, err := client.NeedsUpdate(ctx, registryOpts)
+	if err != nil {
+		return "", xerrors.Errorf("unable to check if built-in policies need to be updated: %w", err)
 	}
 
 	if needsUpdate {
-		log.InfoContext(ctx, "Need to update the built-in checks")
-		log.InfoContext(ctx, "Downloading the built-in checks...")
+		log.InfoContext(ctx, "Need to update the checks bundle")
+		log.InfoContext(ctx, "Downloading the checks bundle...")
 		if err = client.DownloadBuiltinChecks(ctx, registryOpts); err != nil {
-			return nil, xerrors.Errorf("failed to download built-in policies: %w", err)
+			return "", xerrors.Errorf("failed to download checks bundle: %w", err)
```

If a user already has checks in the cache and they have failed to update, should they use them rather than fall back to the embedded ones?

---

Good question, today it would depend on the value of `--skip-check-update`. 

If `--skip-check-update` was not passed in, user already has a cache, but failed to update, in this case we [fall back](https://github.com/aquasecurity/trivy/blob/d0e8626a1d1e8914f1eae7059736c55ea4a0feb7/pkg/commands/artifact/run.go#L649) to embedded. 

If `--skip-check-update` was passed in, user already has a cache, but failed to update, in this case we would use the cache.

I could argue that in both cases the behaviour should be so that we fallback to the embedded checks as an update failure can be due to a corrupt cache. But then again we have [this](https://github.com/aquasecurity/trivy/blame/d0e8626a1d1e8914f1eae7059736c55ea4a0feb7/pkg/commands/artifact/run.go#L649) condition which now I'm not sure serves what purpose.

---

You can remove [this](https://github.com/aquasecurity/trivy/blob/d0e8626a1d1e8914f1eae7059736c55ea4a0feb7/pkg/commands/artifact/run.go#L649) condition and log the error always. Then we can simplify the error message [here](https://github.com/aquasecurity/trivy/pull/8533/files#diff-933179b3c56a9d292175c912e755a45fa9221492b5bdb0c377c798d2766af2ffR93) so that we don't have to duplicate information about fallback to embedded checks.

---

Right, an error when updating checks can occur in the following steps:
- Reading metadata from the cache
- Loading the bundle
- Updating metadata

But maybe we can leave it as it is to simplify the logic.

---

Updated https://github.com/aquasecurity/trivy/pull/8533/commits/56f99c1e55c52cdcbf619fd95bdd01e4f9b91e65